File: descrete/lab02/relation.py
"""
关系性质判定 & 闭包运算（Python 版）
"""
from copy import deepcopy


# —— 性质判定 ——
def is_reflexive(m):
    n = len(m)
    for i in range(n):
        if m[i][i] != 1:
            return False
    return True


def is_irreflexive(m):
    n = len(m)
    for i in range(n):
        if m[i][i] != 0:
            return False
    return True


def is_symmetric(m):
    n = len(m)
    for i in range(n):
        for j in range(n):
            if m[i][j] != m[j][i]:
                return False
    return True


def is_anti_symmetric(m):
    n = len(m)
    for i in range(n):
        for j in range(n):
            if i != j and m[i][j] == 1 and m[j][i] == 1:
                return False
    return True


def is_transitive(m):
    n = len(m)
    for i in range(n):
        for j in range(n):
            if m[i][j] == 1:
                for k in range(n):
                    if m[j][k] == 1 and m[i][k] != 1:
                        return False
    return True


# —— 闭包运算 ——
def reflexive_closure(m):
    c = deepcopy(m)
    for i in range(len(c)):
        c[i][i] = 1
    return c


def symmetric_closure(m):
    c = deepcopy(m)
    n = len(c)
    for i in range(n):
        for j in range(n):
            if c[i][j] == 1:
                c[j][i] = 1
    return c


def transitive_closure(m):
    c = deepcopy(m)
    n = len(c)
    for k in range(n):
        for i in range(n):
            for j in range(n):
                c[i][j] = c[i][j] or (c[i][k] and c[k][j])
    return c


def warshall_steps(m):
    """
    返回每一步矩阵（含初始），用于前端动画
    """
    R = deepcopy(m)
    n = len(R)
    steps = [deepcopy(R)]
    for k in range(n):
        for i in range(n):
            for j in range(n):
                R[i][j] = R[i][j] or (R[i][k] and R[k][j])
        steps.append(deepcopy(R))
    return steps

File: frontend/src/utils/relation.js
// 五种性质判定
export function isReflexive(m)   { const n=m.length; for(let i=0;i<n;i++) if(m[i][i]!==1) return false; return true }
export function isIrreflexive(m) { const n=m.length; for(let i=0;i<n;i++) if(m[i][i]!==0) return false; return true }
export function isSymmetric(m)   { const n=m.length; for(let i=0;i<n;i++) for(let j=0;j<n;j++) if(m[i][j]!==m[j][i]) return false; return true }
export function isAntiSymmetric(m){ const n=m.length; for(let i=0;i<n;i++) for(let j=0;j<n;j++) if(i!==j && m[i][j]===1 && m[j][i]===1) return false; return true }
export function isTransitive(m)  { const n=m.length; for(let i=0;i<n;i++) for(let j=0;j<n;j++) if(m[i][j]===1) for(let k=0;k<n;k++) if(m[j][k]===1 && m[i][k]!==1) return false; return true }

// 闭包
export function reflexiveClosure(m)  { const c=m.map(r=>r.slice()); for(let i=0;i<c.length;i++) c[i][i]=1; return c }
export function symmetricClosure(m)  { const c=m.map(r=>r.slice()); for(let i=0;i<c.length;i++) for(let j=0;j<c.length;j++) if(c[i][j]===1) c[j][i]=1; return c }
export function transitiveClosure(m) { const c=m.map(r=>r.slice()), n=c.length; for(let k=0;k<n;k++) for(let i=0;i<n;i++) for(let j=0;j<n;j++) if(c[i][k]&&c[k][j]) c[i][j]=1; return c }

// Warshall 逐步记录（用于演示）
export function warshallSteps(m){
  const R = m.map(r=>r.slice()), n=R.length, steps=[JSON.parse(JSON.stringify(R))]
  for(let k=0;k<n;k++){
    for(let i=0;i<n;i++){
      for(let j=0;j<n;j++){
        R[i][j] = R[i][j] || (R[i][k] && R[k][j])
      }
    }
    steps.push(JSON.parse(JSON.stringify(R)))
  }
  return steps
}